src/System.Slices/System/Buffers/OwnedMemory.cs

```diff
@@ -198,21 +207,41 @@ internal interface IMemory<T> : IKnown
         unsafe bool TryGetPointer(long id, out void* pointer);
     }
 
-    public struct DisposableReservation : IDisposable
+    public struct DisposableReservation<T> : IDisposable
     {
-        IKnown _owner;
+        OwnedMemory<T> _owner;
         long _id;
 
-        internal DisposableReservation(IKnown owner, long id)
+        internal DisposableReservation(OwnedMemory<T> owner, long id)
         {
             _id = id;
             _owner = owner;
-            _owner.AddReference(_id);
+            switch(OwnedMemorySettings.s_mode) {
+                case Mode.RefCount:
+                    ((IKnown)_owner).AddReference(_id);
+                    break;
+                case Mode.LocalRefCount:
+                    ReferenceCounter.AddReference(_owner);
+                    break;
+                case Mode.None:
+                    break;
+            }
         }
 
+        public Span<T> Span => _owner.Span;
```

Comment:
Sorry, I just noticed this: Could you at least for now not remove this but instead rename it to DangerousGetSpan().
Even if we settle on refcounting option being the default, we want to have the ability to get span without refcounting in cases where the caller knows that it is safe, e.g. the caller owns the memory lifetime or it is GCed memory.